qubes/storage/zfs.py

```diff
@@ -0,0 +1,2281 @@
+                        lock=False,
+                    )
+
+                for snapshot in snapshots:
+                    # For each snapshot, we're going to get all the clones
+                    # made out of it.  The intention is to assign the snapshot
+                    # to the first clone, so that then the prior owner can
+                    # be deleted (if that's what is being requested).
+                    clones = await self.get_snapshot_clones_async(
+                        snapshot,
+                        log=log,
+                    )
+                    for clone in clones:
+                        if clone in already_promoted:
+                            # Never promote something twice.  It is an error.
+                            continue
+
+                        # We will now promote this clone.  This will make the
+                        # snapshot "owned" by the clone, renaming the snapshot
+                        # accordingly.
+                        log.debug(
+                            "Promoting %s so we can free up %s for removal",
+                            clone,
+                            snapshot,
+                        )
+                        await zfs_async("promote", clone, log=log)
+                        self.cache.invalidate(clone, "snapshots")
+                        self.cache.invalidate(snapshot.volume, "snapshots")
+                        # Remember that we promoted this clone now, so we
+                        # do not attempt to promote it again.
+                        already_promoted.add(clone)
+                        # Only promote the first.  We do not need to promote
+                        # other clones because it is sufficient to promote one
+                        # clone in order for it to "own" the snapshot.
+                        break
+
+                if isinstance(volume_or_snapshot, Volume):
+                    # Retry the removal now.
+                    await self.remove_volume_async(
+                        volume_or_snapshot,
+                        log=log,
+                        lock=False,
+                    )
+                    # There is no else clause for VolumeSnapshot because
+                    # once a snapshot's clone has been promoted, then
+                    # the snapshot is renamed to be "nested under" the
+                    # clone, effectively "removing" the snapshot as far
+                    # as the caller is concerned.
+
+        async with self.cache.lock:
+            # Retry up to 20 times if the dataset is busy.
+            # Sometimes users of the volume return before the
+            # kernel has actually closed the device file.  We must
+            # wait until all users of the device file have closed it.
+            await retry_async(
+                _remove_with_promote_async,
+                DatasetBusy,
+                20,
+                0.25,
+            )
```

Comment:
Q: What file system object should I monitor for changes?

A: There are none.  This is an internal ZFS operation only.  Dataset might be busy because a VM has the device file open, or it may be busy because a clone of the dataset is busy — there are a variety of reasons why, but either way, inotify is going to tell us nothing about that.